### ISSUE TITLE
Uncaught Error: runtime error: cannot internalize time.Time from undefined, must be Date

### ISSUE THREAD

**Issue by oskca** [gopherjs/gopherjs#433]
I'm trying to do a [gopherjs binding for VueJS](https://github.com/oskca/gopherjs-vue ). I found a clear way to do bidirectional data bindings from gopherjs to VueJS and so far it works great :smile:  But I came into this problem when `VueJS` trying to set `time.Time` from javascript side to gopherjs code, error code is here: https://github.com/oskca/gopherjs-vue/blob/master/examples/unwritableTime/main.go#L20. (you can try it with `gopherjs serve`)

The error stack goes to `$internalize`,  : 
```Go
  var timePkg = $packages["time"];
  if (timePkg !== undefined && t === timePkg.Time) {
    if (!(v !== null && v !== undefined && v.constructor === Date)) {
      $throwRuntimeError("cannot internalize time.Time from " + typeof v + ", must be Date");
    }
    return timePkg.Unix(new $Int64(0, 0), new $Int64(0, v.getTime() * 1000000));
  }
```

Normal gopherjs code with `time.Time` works fine, but in this case `VueJS` probably tries to set `time.Time` with a function or some other customed data types.

How could this problem  be solved?  Isn't `!(v !== null && v !== undefined && v.constructor === Date)` too much restricted? what about just return `time.Time{}` instead?

**Comment by oskca** [comment]
@neelance At first I thougt this error relates to VueJS, but the following example does not work ether:
```go
package main

import (
	"github.com/gopherjs/gopherjs/js"
	"time"
)

type Test struct {
	*js.Object
	Time time.Time `js:"Time"`
	Text string    `js:"Text"`
}

func main() {
	t := &Test{
		Object: js.Global.Get("Object").New(),
	}
	t.Text = "Hello World"
	println(time.Now())   // this works
	t.Time = time.Now() // this line would panic the program
}
```

Is `time.Time` field not supported yet?

**Comment by shurcooL** [comment]
This is not a bother, thanks for reporting it. I think this is likely a valid issue.

I plan to get around to looking at this soon, I just haven't had a chance yet.

Internalizing/externalizing `time.Time` had some issues which I worked on resolving in the past, see #287. It might be that there's still additional things to resolve.

Can you check which version of `gopherjs` compiler you're using, is it the latest `master`? Also, what does `go version` print?

**Comment by oskca** [comment]
Yes, I'm using the latest `gopherjs` master 64858154c27789453689ac91, my `go version` outputs:
`go1.6 linux/amd64`

**Comment by shurcooL** [comment]
Great, thanks for confirming.

```Go
type Test struct {
    *js.Object
    Time time.Time `js:"Time"`
    Text string    `js:"Text"`
}
```

The entire internalization/externalization logic between Go's `time.Time` and JavaScript's `Date` is quite custom and specific.

I suspect the problem is that it's simply not supported well when you try to use in combination with the JS field tag like ``js:"Time"``. At least that's my initial guess, but we'll want to look more closely at this.

**Comment by shurcooL** [comment]
I should also confirm, are you aware that Go's `time.Time` maps to JavaScript's `Date` object? It can be seen here:

https://godoc.org/github.com/gopherjs/gopherjs/js

Here's a test that exercises some of that:

https://github.com/gopherjs/gopherjs/blob/master/tests/testdata/time_inexternalization.go

> But I came into this problem when VueJS trying to set time.Time from javascript side to gopherjs code

Are you sure VueJS has a `Date` object field and not something else?

**Comment by oskca** [comment]
Yes. I'm aware of the Date/time.Time conversion, but for VueJS I've only read the documentations. And I think VueJS would not pass a Date object directly to the underly laying gopherjs code.

**Comment by oskca** [comment]
But for the  simpler example here: https://github.com/oskca/gopherjs-vue/blob/master/examples/unwritableTime/main.go#L20 with no VueJS code, It still show the `Uncaught Error`.

**Comment by oskca** [comment]
I'm currently working on [gopherjs-vue](https://github.com/oskca/gopherjs-vue) which expose golang struct to VueJS for bidirectional data bindings. Whit he help of `js.MakeWrapper` I can expose the `struct method function` the same time to VueJS as event handlers, which is pretty cool, and thank you again for your great gopherjs project :+1: 

Looking into the `internalization/externalization` code, it seems that `externalization` is for passing data to JavaScript world, and `internalization` is to setting data back to gopherjs. 

Though VueJS works with `bidirectional` data binding, some times we only need to pass data from gopherjs to JavaScript (for imformation displaying), setting back is not always reqired. Since gopherjs  `externalization` procedure can convert (simple or complex) golang data types  correctly to JavaScript , could the `internalization` process be make more robust (ie: if the data type mismatch happens, just set gopherjs value to the default empty value and print WARNING instead of panic?)

**Comment by oskca** [comment]
BTW: it seems that struct with no `*js.Object` can not be `internalized` with gopherjs `$internalize`?

**Comment by shurcooL** [comment]
> But for the  simpler example here: https://github.com/oskca/gopherjs-vue/blob/master/examples/unwritableTime/main.go#L20 with no VueJS code, It still show the `Uncaught Error`.

Yeah, I agree that's a bug and should be fixed.
